geode-redis/src/main/java/org/apache/geode/redis/internal/data/RedisHash.java

```diff
@@ -203,49 +261,142 @@ public int hstrlen(ByteArrayWrapper field) {
+
+    // returnCursorValue = new BigInteger(String.valueOf(0));
+
+    return new ImmutablePair<>(returnCursorValue, resultList);
+  }
+
+  private void removeHSCANSnapshot(UUID clientID) {
+    this.hScanSnapShots.remove(clientID);
+    this.hScanSnapShotCreationTimes.remove(clientID);
+
+    if (this.hScanSnapShots.isEmpty()) {
+      shutDownHscanSnapshotScheduledRemoval();
+    }
+  }
+
+  private List<Object> getResultList(Pattern matchPattern, int count,
+      Map<ByteArrayWrapper, ByteArrayWrapper> dataForKeySnapShot) {
+
+    int currentCursor = 0;
+    List<Object> resultList = new ArrayList<>();
+
+    for (Map.Entry<ByteArrayWrapper, ByteArrayWrapper> entry : dataForKeySnapShot.entrySet()) {
+
+      if (count == currentCursor) {
+        break;
+      }
+
+      currentCursor++;
+
       ByteArrayWrapper key = entry.getKey();
       ByteArrayWrapper value = entry.getValue();
-      i++;
-      if (beforeCursor.compareTo(cursor) < 0) {
-        beforeCursor = beforeCursor.add(new BigInteger("1"));
-        continue;
-      }
 
       if (matchPattern != null) {
-        if (matchPattern.matcher(key.toString()).matches()) {
-          returnList.add(key);
-          returnList.add(value);
-          numElements++;
-        }
+        addMatches(matchPattern, resultList, key, value);
       } else {
-        returnList.add(key);
-        returnList.add(value);
-        numElements++;
+        resultList.add(key);
+        resultList.add(value);
       }
+    }
+    resultList.add(new ByteArrayWrapper(String.valueOf(currentCursor).getBytes()));
+    return resultList;
+  }
 
-      if (numElements == count) {
-        break;
-      }
+  private void addMatches(Pattern matchPattern, List<Object> returnList,
+      ByteArrayWrapper key,
+      ByteArrayWrapper value) {
+    if (matchPattern.matcher(key.toString()).matches()) {
+      returnList.add(key);
+      returnList.add(value);
     }
+  }
 
-    Pair<BigInteger, List<Object>> scanResult;
-    if (i >= size - 1) {
-      scanResult = new ImmutablePair<>(new BigInteger("0"), returnList);
-    } else {
-      scanResult = new ImmutablePair<>(new BigInteger(String.valueOf(i + 1)), returnList);
+  @SuppressWarnings("unchecked")
+  private Map<ByteArrayWrapper, ByteArrayWrapper> getDataForKeys(
+      List<ByteArrayWrapper> keysForScan) {
+    Map<ByteArrayWrapper, ByteArrayWrapper> dataMap = new HashMap<>();
+
+    keysForScan.forEach(key -> {
+      ByteArrayWrapper value = this.hash.get(key);
+      dataMap.put(key, value);
+    });
+
+    return dataMap;
+  }
+
+  // either increment the cursor by the num,ber iterations we did, or
+  // if we're at the end, we return 0
+  // numberOfIterations < count is assumed, is ther a better way?
+
+  private BigInteger getCursorValueToReturn(int count, int startCursor,
+      int numberOfIterationsCompleted) {
+
+    if (numberOfIterationsCompleted < count) {
+      return BigInteger.valueOf(0);
```

Comment:
why not return an integer here that the caller can then create a BigInteger from if that's what they need?

Reply:
done